amads/melody/contour/interpolation_contour.py

```diff
@@ -0,0 +1,272 @@
+        https://www.doc.gold.ac.uk/isms/m4s/FANTASTIC_docs.pdf
+        """
+        # Step 1 & 2: Find contour extrema
+        extrema_indices = self._find_contour_extrema(self.pitches)
+        extrema_times = [self.times[i] for i in extrema_indices]
+        extrema_pitches = [self.pitches[i] for i in extrema_indices]
+
+        # Step 3: Calculate gradients
+        gradients = np.diff(extrema_pitches) / np.diff(extrema_times)
+
+        # Step 4: Calculate durations
+        durations = np.diff(extrema_times)
+
+        # Step 5: Remove durations below 0.05 seconds
+        durations[durations < 0.05] = 0
+
+        # Step 6: Create weighted gradients vector by repeating each gradient
+        # proportional to its duration
+        # Convert durations to integer number of samples (at 100Hz sampling rate)
+        samples_per_duration = np.round(durations * 100).astype(int)
+        interpolation_contour = np.repeat(gradients, samples_per_duration)
+        return [float(x) for x in interpolation_contour]
+
+    @property
+    def direction(self) -> list[int]:
+        """Calculate the global direction of the interpolation contour.
+
+        Returns
+        -------
+        list[int]
+            List containing the signs (+1, 0, -1) of the gradient values in the contour
+
+        Examples
+        --------
+        >>> ic = InterpolationContour([0, 0.1, 0.2, 0.3], [60, 58, 64, 65])
+        >>> ic.direction[:4]  # Show first 4 values
+        [-1, -1, -1, -1]
+        """
+        # Calculate the signum of the interpolation contour using numpy's sign function
+        global_direction = np.sign(self.contour)
+
+        return [int(x) for x in global_direction]
+
+    @property
+    def mean_gradient(self) -> float:
+        """Calculate the absolute mean gradient of the interpolation contour.
+
+        Returns
+        -------
+        float
+            Mean of the absolute gradient values
+
+        Examples
+        --------
+        >>> ic = InterpolationContour([0, 0.1, 0.2, 0.3], [60, 58, 64, 65])
+        >>> ic.mean_gradient
+        30.0
+        """
+        return float(np.mean(np.abs(self.contour)))
+
+    @property
+    def gradient_std(self) -> float:
+        """Calculate the standard deviation of the interpolation contour gradients.
+
+        Returns
+        -------
+        float
+            Standard deviation of the gradient values
+
+        Examples
+        --------
+        >>> ic = InterpolationContour([0, 0.1, 0.2, 0.3], [60, 58, 64, 65])
+        >>> ic.gradient_std
+        25.9272...
+        """
+        return float(np.std(self.contour))
+
+    @property
+    def direction_change_ratio(self) -> float:
+        """Calculate the ratio of direction changes in the interpolation contour.
+
+        This measures the number of changes in contour direction relative to the number
+        of interpolation lines (i.e. number of different gradient values).
+
```

Comment:
Could be worth explaining a bit more what the intuition is behind this.

Reply:
I regret that Müllensiefen doesn't actually discuss why this might be useful in the tech report - I suppose it tells us something about global direction variability?